### PR TITLE
Scaffold server config via a bootstrapping script

### PR DESCRIPTION
**Pull request by shtlrs**
**TL;DR**

Removes the the needed manual work to fill the `config.yaml` file with all the channel/role/category values of the test server

# Pain point
Configuring the bot is a very painful procedure due to how overwhelming & big the configuration file is. This makes contribution to people who aren't familiar with our project very discouraging, and people would often find themselves too lazy to configure it, confused in terms of what to configure and what not, or unable to configure since the contribution guide needs to be read thouroughly, etc.

# Idea
The idea is to have a script that would **magically** build the necessary configuration with a click of a button/a single command.
Since the bot doesn't need all the channels, roles, categories, icons, emojis, etc. to be configured, the script would only prepare the necessary server related variables

This is simply achievable by following these steps
1. Create a test server out of our test server's [template](https://discord.new/zmHtscpYN9E3)
2. [Configure](#configuring-the-script) the bootstrapping script 
3. [Run](#running-the-script) the script
4. Run the bot

> Configuring & launching the site project is out of scope of this tutorial but necessary for the bot to work. Details on how to configure it can be found [here](https://www.pythondiscord.com/pages/guides/pydis-guides/contributing/site/) 

# Configuring the script
This is as simple as
1. Making sure there's a `.env` file created in the root directory of your project (if not, please add it)
2. Setting the `BOT_TOKEN` variable with a value likewise: `BOT_TOKEN=your-bot-token`
3. Setting the `GUILD_ID` variable with a value likewise: `GUILD_ID=your-discord-server-id`

# Running the script
Once `BOT_TOKEN` and `GUILD_ID` are configured, it should be as simple as running `python bootstrap_config.py`

The script will
1. Fetch all channels, categories and roles in the server
2. Save data in a dictionaries where keys represent the channel name, category name or role name, and the value will be the id of that item
3. Iterate the fields in the constants classes that are related to the server, namely `Categories`, `Channels` and `Categories`
6. Try to find a match for each field of the class constants in the previously populated dictionaries.
7. Save the matched ids in the `.env.server` environment variables file

# Changelog
1. Replace the usage of `.yaml` files for configuration with `.env` files
8. Use `Pydantic` for parsing values from the environment variables **AND** the environment files

Todo:
- [ ] Add checks on mandatory variables that need to be present
- [ ] Update guide on site
- [ ] Discuss backwards compatibility solutions (if supported)
- [ ] Renew the template so that warnings of missing channels/roles/categories/etc. reduce down

### CODE REVIEW

File: bot/exts/filters/antispam.py [hunk 1]
```diff
@@ -297,10 +297,13 @@ async def on_message_edit(self, before: Message, after: Message) -> None:
         self.cache.update(after)
 
 
-def validate_config(rules_: Mapping = AntiSpamConfig.rules) -> Dict[str, str]:
+def validate_config(rules_: Mapping = None) -> Dict[str, str]:
```
Comment (HassanAbouelela) [inline bot/exts/filters/antispam.py:300]: Is it not possible to move the `AntiSpamConfig.rules.dict()` as the default value, similar to what was there before?

**Review by HassanAbouelela — CHANGES_REQUESTED** [review]
Looks good overall. Some design nitpicks I would like addresed though.

File: bootstrap_config.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        "Couldn't find GUILD_ID in the environment variables."
+        "Make sure to add it to the `.env` file likewise: `GUILD_ID=value_of_your_discord_server_id`"
+    )
+    log.warning(message)
+    raise ValueError(message)
+
+
+BASE_URL = "https://discord.com/api/v10"
+
+HEADERS = {"Authorization": f"Bot {BOT_TOKEN}"}
```
[the remark this answers is not in the document]
Comment (shtlrs) [reply bootstrap_config.py:37]: If I'm not mistaken, `Session` doesn't offer a base url feature, I'll extend the `Session` class to alter the url upon each request

File: bootstrap_config.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+
+def get_all_roles(guild_id: int | str) -> dict:
+    """Fetches all the roles in a guild."""
+    result = {}
+
+    roles_url = f"{BASE_URL}/guilds/{guild_id}/roles"
+    response = requests.get(url=roles_url, headers=HEADERS)
```
[the remark this answers is not in the document]
Comment (shtlrs) [reply bootstrap_config.py:45]: Will do, however we wouldn't want to do that for `get_webhook` since we need to return the empty object for us to know that it doesn't exist. Or just catch it there

File: bootstrap_config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (HassanAbouelela) [reply bootstrap_config.py:37]: I believe you are correct, hence the recommendation for httpx. It’s generally more powerful

File: bootstrap_config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (HassanAbouelela) [reply bootstrap_config.py:45]: That sounds fine (side note: should we return a more-explicit null?)

[the remark this answers is not in the document]
Comment (shtlrs) [reply bootstrap_config.py:45]: I actually was planning on making it a predicate, so basically a yes or no question to whether the webhook is there or not

File: bootstrap_config.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
+import os
+from pathlib import Path
+
+import requests
```
[the remark this answers is not in the document]
Comment (shtlrs) [reply bootstrap_config.py]: Done here 42741ef1f41bba8ad534b23b8f4cb5b2076b83d8

File: bootstrap_config.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shtlrs) [reply bootstrap_config.py]: Done in 887e88e4f4c6bcde3dbe1e956d549a01fb669820 and 8408f4dd4824ceb6c8902bc613e60462eb13d78f

File: bootstrap_config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shtlrs) [reply bootstrap_config.py]: Done here 0c9cd82016ee9b67f02d12f8874cfd33fdc74570

File: bot/constants.py [hunk 5]
```diff
@@ -10,694 +10,728 @@
 their default values from `config-default.yml`.
```
[the remark this answers is not in the document]
Comment (shtlrs) [reply bot/constants.py]: Updated in e5cfc9e2c7378cc5041865dc877e1f15115edc7f

File: bot/exts/filters/antispam.py (hunk 1 above)
Comment (shtlrs) [reply bot/exts/filters/antispam.py:300]: That's what I did in the first place, but flake8 didn't like that.
Here's what it says

```error
B008 Do not perform function calls in argument defaults.  The call is performed only once at function definition time. All calls to your function will reuse the result of that definition-time fun
ction call.  If this is intended, assign the function call to a module-level variable and use that variable as a default value.
```
Which is why I did that, I can either leave it the way it is or tell flake8 to ignore it, WDYT ?

File: bot/constants.py [hunk 6]
```diff
@@ -10,694 +10,728 @@
...
+        env_file_encoding = 'utf-8'
 
-        # Optional app configuration. Set `MY_APP_KEY` in the environment to use it.
-        application:
-            key: !ENV 'MY_APP_KEY'
-    """
 
-    default = None
+class _Miscellaneous(EnvConfig):
+    debug: str = Field(env="BOT_DEBUG", default="true")
```
[the remark this answers is not in the document]
Comment (shtlrs) [reply bot/constants.py:32]: That's not actually it. This is there to ensure backwards compatibility with the env variables that might be in production.
The values are not the same, as you can see here we don't define a prefix for this specific class, unlike the rest where it's usually `something__`

Since in the old setting this was extracted from a pure env variable called BOT_DEBUG, I thought I'd leave it the way it is.
But if we were to change it, we need to make sure that we define them in the prod config as well. WDYT ?

File: bootstrap_config.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+env_file_path.write_text(config_str)
+
+config_str += "\n#Webhooks\n"
+
+
+for webhook_name, webhook_model in Webhooks:
+    webhook = get_webhook(webhook_model.id)
+    if not webhook:
+        webhook_id = create_webhook(webhook_name, webhook_model.channel)
```
[the remark this answers is not in the document]
Comment (shtlrs) [reply bootstrap_config.py]: That's true, my bad.
I had changed it before & then I wanted to experiment further with something else.
Fixed here 97a5d864e597441cd7f19c8e2dc53d919af64448

File: bot/exts/filters/antispam.py (hunk 1 above)
Comment (shtlrs) [reply bot/exts/filters/antispam.py]: Well, I ended up making it a module level var here db665dcd1ccdd89ff4bd1c20bfb3c340418c86c3

File: bot/constants.py [hunk 8]
```diff
@@ -10,694 +10,728 @@
...
+from pydantic import BaseModel, BaseSettings, Field, root_validator
 
-try:
-    import dotenv
-    dotenv.load_dotenv()
-except ModuleNotFoundError:
-    pass
+# Will add a check for the required keys
 
+env_file_path = Path(__file__).parent.parent / ".env"
```
[the remark this answers is not in the document]
Comment (shtlrs) [reply bot/constants.py]: Done in 5ee1bfd5006ffd63e220de72c971a279916f18fe

File: bot/constants.py [hunk 9]
```diff
@@ -10,694 +10,728 @@
...
-    the document to strings. This allows for a much more DRY configuration
-    file.
-    """
 
-    fields = loader.construct_sequence(node)
-    return "".join(str(x) for x in fields)
+class _Bot(EnvConfig):
+    EnvConfig.Config.env_prefix = "bot__"
 
+    prefix: str = Field(default="!")
```
[the remark this answers is not in the document]
Comment (shtlrs) [reply bot/constants.py:45]: Do you want to resolve this and keep discussing it in https://github.com/python-discord/bot/pull/2408#discussion_r1116254111 ?

File: bootstrap_config.py [hunk 10]
```diff
@@ -72,74 +83,76 @@ def get_all_channels_and_categories(guild_id: int | str) -> tuple[dict[str, str]
...
-
-all_roles = get_all_roles(guild_id=GUILD_ID)
-
-for role_name in _Roles.__fields__:
-
-    role_id = all_roles.get(role_name, None)
-    if not role_id:
-        log.warning(f"Couldn't find the role {role_name} in the guild, PyDis' default values will be used.")
-        continue
+if __name__ == '__main__':
```
Comment (HassanAbouelela) [inline bootstrap_config.py:104]: I don't think we should have this guard. This file should in theory not be imported anywhere, and it adds an unnecessary layer of indentation.

File: bot/constants.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (HassanAbouelela) [reply bot/constants.py:26]: I've accepted the upfront migration as a cost of the upgrade. We should not add confusion/maintenance cost to avoid 10 minutes of manual work. This also applies to nearly every single Field in this file.

File: bot/constants.py [hunk 11]
```diff
@@ -563,6 +556,7 @@ class _URLs(_BaseURLs):
 
     site_staff: str = "".join([BaseURLs.site_schema, BaseURLs.site, "/staff"])
     site_paste = "".join(["paste.", BaseURLs.site])
+    site_api = "".join([BaseURLs.site, "/api"])
```
Comment (HassanAbouelela) [inline bot/constants.py:559]: Could you clarify this change? This would not have the correct value in production and require overriding. I prefer to remain consistent and just have prod values, and require devs override.

**Review by HassanAbouelela — CHANGES_REQUESTED** [review]

File: bootstrap_config.py (hunk 10 above)
Comment (shtlrs) [reply bootstrap_config.py]: removed in 469d609ce53a3e7e1378672c4787cc80444450ba

File: bot/constants.py (hunk 11 above)
Comment (shtlrs) [reply bot/constants.py:559]: I'm not sure I understand what you mean by [this](https://github.com/python-discord/bot/pull/2408#discussion_r1117955608)

Are you talking about the name of the class itself and having its own prefix ?
Or are you talking about passing the `env` kwarg to specify how to fetch it from env variables ?
Or none of the above ?

**Comment by shtlrs** [comment]
@HassanAbouelela  I'm not sure I understand what you mean by [this](https://github.com/python-discord/bot/pull/2408#discussion_r1117955608)

Are you talking about the name of the class itself and having its own prefix ?
Or are you talking about passing the `env` kwarg to specify how to fetch it from env variables ?
Or none of the above ?

File: bot/constants.py (hunk 11 above)
Comment (shtlrs) [reply bot/constants.py]: Reverted in 686590c0d797aea28d79d94ea939277f734f404d

I got confused & misunderstood something while I was updating the docs. I think my brains got mangled with all of the renaming of the configs & I ticked in my head that we were constructing the site_api in that fashion.

My bad..

**Comment by HassanAbouelela** [comment]
I'm referring to creating Fields and overriding pydantic's default env key creation. I.e, in the line commented on, the diff would be:
```diff
- debug: str = Field(env="BOT_DEBUG", default="true")
+ debug = True
```

And in prod we'll update the `BOT_DEBUG` key to whatever pydantic resolved it as.

**Comment by shtlrs** [comment]
> I'm referring to creating Fields and overriding pydantic's default env key creation. I.e, in the line commented on, the diff would be:
> 
> ```diff
> - debug: str = Field(env="BOT_DEBUG", default="true")
> + debug = True
> ```
> 
> And in prod we'll update the `BOT_DEBUG` key to whatever pydantic resolved it as.

The confusion happened due to the lack of knowledge that we were ready to update the env var keys in prod as well.

Done in 36ebda1d54cb88033e4be2ef399668f0c3664a15

**Review by HassanAbouelela — APPROVED** [review]

File: bot/constants.py [hunk 12]
```diff
@@ -1,703 +1,728 @@
...
+
+
+class _BaseURLs(EnvConfig):
+    EnvConfig.Config.env_prefix = "urls."
 
     # Snekbox endpoints
-    snekbox_eval_api: str
-    snekbox_311_eval_api: str
+    snekbox_eval_api = "http://snekbox.default.svc.cluster.local/eval"
+    snekbox_311_eval_api = "http://snekbox-311.default.svc.cluster.local/eval"
```
Comment (HassanAbouelela) [inline bot/constants.py:528]: These are out of date, new values are:
snekbox_eval_api: `http://snekbox-310.default.svc.cluster.local/eval`
snekbox_311_eval_api: `http://snekbox.default.svc.cluster.local/eval`

**Review by HassanAbouelela — COMMENTED** [review]
Just missed this.

**Comment by HassanAbouelela** [comment]
New environment variables deployed under `new-bot-env`, ready to be migrated when we merge this PR.

File: bot/constants.py (hunk 12 above)
Comment (shtlrs) [reply bot/constants.py]: Tadaa a839fa59d9fae8151401e309eda29dc680503bed

**Comment by HassanAbouelela** [comment]
Hey there's one thing I forgot, the current bootstrap doesn't handle off-topic channels which automatically change names every 24 hours. Can we special case OT channels during bootsrap? Generally, the names will always take the form of `ot<int>-name...`, and be in the Off-Topic category. If you have a preferred implementation, please proceed with it, otherwise let me know.

File: bot/constants.py [hunk 13]
```diff
@@ -1,703 +1,728 @@
...
+    pyweek_announcements = 897568414044938310
+    revival_of_code = 988801794668908655
+    legacy_help_channels_access = 1074780483776417964
+
+    contributors = 295488872404484098
+    help_cooldown = 699189276025421825
+    muted = 277914926603829249
+    partners = 323426753857191936
+    python_community = 458226413825294336
+    sprinters = 758422482289426471
```
Comment (HassanAbouelela) [inline bot/constants.py:149]: Let's drop this. it's only used for a (now unused) whitelist. If you stumble upon any unused constants like this, feel free to drop as well.

**Comment by shtlrs** [comment]
> Hey there's one thing I forgot, the current bootstrap doesn't handle off-topic channels which automatically change names every 24 hours. Can we special case OT channels during bootsrap? Generally, the names will always take the form of `ot<int>-name...`, and be in the Off-Topic category. If you have a preferred implementation, please proceed with it, otherwise let me know.

It's been handled in 4b4b65bfe5b923e9c2ac6f6cc137210fa936ccdf

**Review by HassanAbouelela — APPROVED** [review]

File: botstrap.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+    log.warning(message)
+    raise ValueError(message)
+
+
+class DiscordClient(Client):
+    """An HTTP client to communicate with Discord's APIs."""
+
+    def __init__(self):
+        super().__init__(
+            base_url="https://discord.com/api/v10",
```
Comment (onerandomusername) [inline botstrap.py:41]: why is this being provided?

Comment (onerandomusername) [reply botstrap.py:41]: Nevermind, didn't realise this *wasn't* a bot lib subclass.

File: bot/constants.py [hunk 15]
```diff
@@ -1,703 +1,726 @@
...
+
+    # Self-assignable roles, see the Subscribe cog
+    advent_of_code = 518565788744024082
+    announcements = 463658397560995840
+    lovefest = 542431903886606399
+    pyweek_announcements = 897568414044938310
+    revival_of_code = 988801794668908655
+    legacy_help_channels_access = 1074780483776417964
+
+    contributors = 295488872404484098
```
Comment (onerandomusername) [inline bot/constants.py:144]: there should be a comment before these. Right now at first glance it looks as if they're self assignable

**Review by ChrisLovering — APPROVED** [review]
Looks good to me, tested locally and all looks good.

**Comment by HassanAbouelela** [comment]
This is being merged now, handling any issues.
